Test linalg vector norm subgradient and fix nans when p < 1

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74205
* __->__ #75103

This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

soulitzer committed:
Test linalg vector norm subgradient

[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer commented:
@albanD looks like ASAN tests are already failing for backward already when zero is passed, I think it is ok to just silence it

albanD commented:
Yes. It should be a high pri issue to fix it though.

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -190,6 +190,8 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
 }
 
 Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: zero entries of `norm` (and in one case, `self`) are replaced with something
```
What does "one case" means here? Does that refer to the test cases?

albanD reviewed (commented):
cc @ngimel what is your feeling about this? Is the perf hit worth it to appease ASAN build?

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

ngimel commented:
I don't think masked_fill just to appease ASAN is worth it, we are filling those values later anyway with zeros, right?

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"




[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"


This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -189,7 +189,10 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
-Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+Tensor norm_backward(
+    Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: we could've masked_fill the zero elements with an arbitrary value like 1 to
```
I don't think this comment is precise enough. What are you maske filling? And why?

soulitzer replied:
Updated!

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -205,31 +208,35 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    return self_scaled * scale_v;
   } else {
-    self_scaled = self * self.abs().pow(p - 2);
+    self_scaled = self * self.abs().pow_(p - 2);
```
If we're adding inplace, we can make the multiplication inplace as well here.
Here and above

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -205,31 +208,35 @@ Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p
+    return self_scaled * scale_v;
   }
-  // handle case at 0 where we return a subgradient containing 0
-  scale_v.masked_fill_(norm == 0, 0);
```
If there is only one case where you don't do it, then I think it is worth leaving this here and having the one that doesn't need this do an early return?

soulitzer replied:
Hmm there are two actually, the p < 1.0 case masks self and doesn't need to mask norm, so there will a little extra logic to handle that if I move it to below

albanD replied:
Ok then!

albanD reviewed (commented):
Do we have an issue tracking what needs to done

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"


This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"


This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

[ghstack-poisoned]

soulitzer committed:
Update on "Test linalg vector norm subgradient and fix nans when p < 1"


This PR:
- ignores sanitizer for pow kernel when p = -2 (ignore does not apply recursively, so need to apply at the root)
- splits OpInfo so that one tests zeros as inputs (with a single OpInfo we'd need to skip the gradgrad tests)
- corrects a division by zero issue that produces NaNs when p < 1
- avoids an extra unneeded masked fill for the inf norm case

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -189,7 +189,12 @@ Tensor norm_backward(const Tensor& grad, const Tensor& self, const optional<Scal
-Tensor norm_backward(Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+Tensor norm_backward(
+    Tensor grad, const Tensor& self, const optional<Scalar> & p_, Tensor norm, IntArrayRef dim, bool keepdim) {
+  // NB: We mask fill the NaNs in the output to be zero but still do float division
```
nit: this should be past tense no?

soulitzer replied:
hmm... debatable :)

albanD reviewed (approved):
Details on the comment. But good otherwise!

soulitzer commented:
@pytorchbot merge this
